server/src/main/java/org/elasticsearch/health/node/action/TransportHealthNodeAction.java

```diff
@@ -0,0 +1,125 @@
+    HandledTransportAction<Request, Response> {
+
+    private static final Logger logger = LogManager.getLogger(TransportHealthNodeAction.class);
+
+    protected final TransportService transportService;
+    protected final ClusterService clusterService;
+    protected final ThreadPool threadPool;
+    protected final String executor;
+
+    private final Writeable.Reader<Response> responseReader;
+
+    protected TransportHealthNodeAction(
+        String actionName,
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        Writeable.Reader<Request> request,
+        Writeable.Reader<Response> response,
+        String executor
+    ) {
+        super(actionName, true, transportService, actionFilters, request);
+        this.transportService = transportService;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.executor = executor;
+        this.responseReader = response;
+    }
+
+    protected abstract void healthOperation(Task task, Request request, ClusterState state, ActionListener<Response> listener)
+        throws Exception;
+
+    @Override
+    protected void doExecute(Task task, final Request request, ActionListener<Response> listener) {
+        ClusterState state = clusterService.state();
+        logger.trace("starting to process request [{}] with cluster state version [{}]", request, state.version());
+        if (task != null) {
+            request.setParentTask(clusterService.localNode().getId(), task.getId());
+        }
+        if (isTaskCancelled(task)) {
+            listener.onFailure(new TaskCancelledException("Task was cancelled"));
+            return;
+        }
+        try {
+            ClusterState clusterState = clusterService.state();
+            DiscoveryNode healthNode = HealthNode.findHealthNode(clusterState);
+            DiscoveryNode localNode = clusterState.nodes().getLocalNode();
+            if (healthNode == null) {
+                listener.onFailure(new HealthNodeNotDiscoveredException("Health node was null"));
+            } else if (localNode.getId().equals(healthNode.getId())) {
+                threadPool.executor(executor).execute(() -> {
+                    try {
+                        if (isTaskCancelled(task)) {
+                            listener.onFailure(new TaskCancelledException("Task was cancelled"));
+                        } else {
+                            healthOperation(task, request, clusterState, listener);
+                        }
+                    } catch (Exception e) {
+                        listener.onFailure(e);
+                    }
+                });
+            } else {
+                logger.trace("forwarding request [{}] to health node [{}]", actionName, healthNode);
+                transportService.sendRequest(
```

Comment:
I think this should be `sendChildRequest` so that cancellation (and tracing etc) will propagate correctly.

Reply:
Oh right as Andrei points out we are explicitly setting up the parent anyway. Yeah I'd prefer this to use `sendChildRequest` instead of doing that but at least it's not a bug either way.

Reply:
TIL! @DaveCTurner thanks, I will use this if we have a task then, does it mean that the following code is redundant or does it serve a different purpose? It's one of the first things we do in `doExecute(..)`
```
        if (task != null) {
            request.setParentTask(clusterService.localNode().getId(), task.getId());
        }
```

Reply:
Nvm, I saw @andreidan already confirmed that :)!